server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -788,6 +788,43 @@ public void batchDelete(Collection<Entitlement> entitlements) {
         }
     }
 
+    /**
+     * Deletes the entitlements with the given collection of ids, using a SQL delete.
+     *
+     * WARNING: This method does not maintain runtime consistency for any of the entitlements it is deleting,
+     * so the caller should take care of that manually for any of those entitlements that are loaded,
+     * preferably by calling {@link #preDeleteOperations} first.
+     *
+     * @param entitlementIds The collection of ids of the entitlements to be deleted.
+     */
+    public void batchDeleteByIds(Collection<String> entitlementIds) {
+        Map<String, Object> criteria = new HashMap<>();
+        criteria.put("id", entitlementIds);
+        this.bulkSQLDelete(Entitlement.DB_TABLE, criteria);
+    }
+
+    /**
+     * Maintains the runtime consistency of the given collection of entitlements, by removing references to
+     * themselves and their certificates. This method is supposed to be run before performing the deletion
+     * of the given list of entitlements.
+     *
+     * @param entitlements The collection of entitlements whose references are to be removed.
+     */
+    public void preDeleteOperations(Collection<Entitlement> entitlements) {
```

Comment:
How's the memory usage with this step added back in? Are we getting back to the point where this is reintroducing the memory problems?

Reply:
Nope. The memory usage is almost exactly the same as without it. The memory issues where from the hibernate Session itself accumulating data of all the entities that are to be deleted before we flush. The entitlements and their consumers/pools that this method operates on are either already loaded, or will be loaded later on in the deletePools method anyway, so there is no impact in memory from it.

Reply:
Classic Hibernate.